### ISSUE TITLE
Why not just a server!?

### ISSUE THREAD

**Issue by vaibhav-kaushal** [goradd/goradd#140]
Hey @spekary 

I think the shape is good but I also think that the industry is fast moving towards decoupling. Most apps do not render their own HTML a lot. They do the minimal stuff (just the case content) and then let JS do the work on the front end. Most sites that are going to be accompanied by an app these days and thus, they require the APIs and front-end to stay separated. Is there something you are working on towards that?

**Comment by spekary** [comment]
That is a great question. The short answer is that goradd allows a single developer to build at the initial stages of a project, and then expand from there. It does not force the architecture you are describing at the beginning, but does allow that kind of architecture to evovle if the project needs it.

Its mainly a question of how the architecture of a project reflects the priorities of the project. I have had to rescue a few projects that hired companies to blindly build apps the way you described. The result was a mess, because that architecture usually requires a team of people who specialize (database designer, API backend designer, Angular guy, IOS guy, Android guy, deployment guy). They also need to have a very good written spec that will not change. And they need a dedicated project manager to tie it all together. Its very resource heavy and inflexible. If you have all of that, know exactly what you want, and believe that you are going to need massive scalability, then that is a reasonable way to go. Most projects do not fit that.

The goradd approach is to build on a technology that is fast enough for most situations and highly flexible. It prioritizes the productivity of the engineers over squeezing out miniscule performance gains before the application has even been prototyped. 

My tests on a 2GHz Mac laptop show that a goradd app can easily serve 1000 requests per second on one server using server-side rendering, with an average response time of 10-50ms under that load. If that is not enough, it is scalable to multiple machines. That should be plenty of horse-power for most projects. Most business apps do not need intensive javascript interactions. They primarily gather information, process, and report. Why complicate that? Users will not know or care how it is implemented as long as its fast.

Regarding building a common API for both app development and web development, I have  found that to be more of a burden than a blessing. Mobile apps generally have really different requirements and goals than the companion website. I have experimented with generated REST and GraphQL apis, and have found it much more useful to build APIs from scratch so you can limit the API interactions with the database. The code-generated ORM in goradd helps make that easy. A goradd app can simultaneously serve web pages, a REST-like API, and respond to WebSocket requests, all from different tcp ports, but from the same application. If traffic grows, those tasks can be broken apart, or the entire application can be replicated. There are a variety of ways to scale depending on what the bottleneck is. Most projects will never get to that point though. 

Now, given all of that, goradd can still help with projects that want a different architecture. For example, someone who wants to use WebComponents, Polymer, or React will need a way of pushing data to a client, rather than waiting for the client to poll for data. Goradd can use WebSockets or a messaging server (pubhub for example) to push data to a data view on the browser or in a mobile app, and then broadcast the change to the rest of the app. Since Goradd is a server, and not just a scripting language like PHP, it has complete control. With GO's go routines, it can even process data in the background on multiple cores of the computer while doing all of that. I am working on a widget to do the dataview thing. Its basically a way of bringing CoachBase type functionality to any database.

Another reason to not render everything server side is for widgets that are scrolling. If the widget is scrolled, and you redraw its html, you lose the scroll position. There are many ways to solve this problem. Goradd has the ability to push portions of an html object, like attributes or innerText, it can communicate directly with an HTML widget over a WebSocket or messaging server, it can link an html widget with a data view like the one I previously described, or it can communicate through Ajax events to custom javascript attached to an html element. 

Finally, regarding modularity. Modularizing an app should not be done blindly. It should be done understanding the benefits and the costs. The goradd approach is to start with internal modularity, using GO's emphasis on unit testing to build a testable product. And then, have an architecture that allows the developer to break things apart into separate services, or microservices, as the need arises. 

So, to summarize, replicating an Angular approach is not the priority right now. The priority is getting to a functional Web app that can also serve a REST api as quickly and flexibly as possible. That said, goradd's ORM, WebSockets, and ability to serve one app through multiple ports can make goradd a beneficial piece of an architecture like that and that is experiencing server-side bottlenecks.
